magpie/api/management/register/register_utils.py

```diff
@@ -1,50 +1,110 @@
+        If overridden, can be any HTML rendered response.
     """
     if tmp_token.expired():
         str_token = str(tmp_token.token)
-        db_session.delete(tmp_token)
+        request.db.delete(tmp_token)
         ax.raise_http(HTTPGone, content={"token": str_token}, detail=s.TemporaryURL_GET_GoneResponseSchema.description)
     ax.verify_param(tmp_token.operation, is_type=True, param_compare=TokenOperation,
                     param_name="token", http_error=HTTPInternalServerError, msg_on_fail="Invalid token.")
+
+    # if any token handler needs to return a custom response (eg: UI message page), it should override this variable
+    response = None
+
     if tmp_token.operation == TokenOperation.GROUP_ACCEPT_TERMS:
         ax.verify_param(tmp_token.group, not_none=True,
                         http_error=HTTPInternalServerError, msg_on_fail="Invalid token.")
         ax.verify_param(tmp_token.user, not_none=True,
                         http_error=HTTPInternalServerError, msg_on_fail="Invalid token.")
-        uu.assign_user_group(tmp_token.user, tmp_token.group, db_session)
-    if tmp_token.operation == TokenOperation.USER_PASSWORD_RESET:
+        uu.assign_user_group(tmp_token.user, tmp_token.group, request.db)
+
+    elif tmp_token.operation == TokenOperation.USER_PASSWORD_RESET:
         ax.verify_param(tmp_token.user, not_none=True,
                         http_error=HTTPInternalServerError, msg_on_fail="Invalid token.")
-        # TODO: reset procedure
+        # TODO: password reset procedure
         ax.raise_http(HTTPNotImplemented, detail="Not Implemented")
-    if tmp_token.operation == TokenOperation.WEBHOOK_USER_STATUS_ERROR:
+
+    elif tmp_token.operation == TokenOperation.WEBHOOK_USER_STATUS_ERROR:
         ax.verify_param(tmp_token.user, not_none=True,
                         http_error=HTTPInternalServerError, msg_on_fail="Invalid token.")
         webhook_update_error_status(tmp_token.user.user_name)
-    db_session.delete(tmp_token)
+
+    # User Registration Procedure - Step (3): reception of the registration email confirmation URL from the user
+    elif tmp_token.operation == TokenOperation.USER_REGISTRATION_CONFIRM_EMAIL:
+        LOGGER.debug("[User Registration - Step 3] received email validation for user: [%s]", tmp_token.user.user_name)
+        response = handle_user_registration_confirmation(tmp_token, request)
+
+    # User Registration Procedure - Step (4): reception of the administrator approve/decline URL
+    elif tmp_token.operation in [TokenOperation.USER_REGISTRATION_ADMIN_APPROVE,
+                                 TokenOperation.USER_REGISTRATION_ADMIN_DECLINE]:
+        ax.verify_param(ar.has_admin_access(request), is_true=True, with_param=False,
+                        http_error=HTTPForbidden, msg_on_fail=s.HTTPForbiddenResponseSchema.description)
+        LOGGER.debug("[User Registration - Step 4] admin reviewed pending user: [%s]", tmp_token.user.user_name)
```

Comment:
Should we log the admin user approving or declining the user?

Reply:
We can do that. Is only a log entry sufficient. 
That info won't be actively stored anywhere, an will be lost according to how frequently the log rotation occurs for example.

Reply:
I added logs to the sub-operation that handles approve/decline cases.

Reply:
I don't thing that we must store that info in a db. I only saw a use case where something doesn't run smoothly and seeing your beautiful logs, that was the only missing piece to know the full story.